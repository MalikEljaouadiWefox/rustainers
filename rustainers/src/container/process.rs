use serde::{Deserialize, Serialize};

use super::{ContainerId, ContainerStatus};

/// Container process
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct ContainerProcess {
    #[serde(alias = "ID")]
    pub(crate) id: ContainerId,
    pub(crate) names: Names,
    pub(crate) state: ContainerStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Names {
    Name(String),
    List(Vec<String>),
}

impl Names {
    pub fn contains(&self, name: &str) -> bool {
        match self {
            Self::Name(n) => n == name,
            Self::List(ns) => ns.iter().any(|n| n == name),
        }
    }
}

#[cfg(test)]
mod tests {
    use assert2::let_assert;

    use super::*;
    #[test]
    fn should_serde_docker_process() {
        let json_stream = include_str!("../../tests/assets/docker-ps.jsonl");
        let stream =
            serde_json::Deserializer::from_str(json_stream).into_iter::<ContainerProcess>();
        let result = stream.collect::<Result<Vec<_>, _>>();
        let_assert!(Ok(data) = result);
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn should_serde_podman_process() {
        let json = include_str!("../../tests/assets/podman_ps.json");
        let result = serde_json::from_str::<Vec<ContainerProcess>>(json);
        let_assert!(Ok(data) = result);
        insta::assert_debug_snapshot!(data);
    }
}
