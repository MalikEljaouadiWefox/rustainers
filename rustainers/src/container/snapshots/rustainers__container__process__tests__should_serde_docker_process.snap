---
source: rustainers/src/container/process.rs
expression: data
---
[
    ContainerProcess {
        id: ContainerId(
            "efb522e2ce2be08d3ff96685c07dd6521c27ddb2aac2c33761f292a74dd872d0",
        ),
        names: Name(
            "redpanda-console",
        ),
        state: Exited,
    },
    ContainerProcess {
        id: ContainerId(
            "58e4e6921c09b9ad17e62ae6fa92455bbe69eca31782b7b848b3972b86af1eb7",
        ),
        names: Name(
            "redpanda-0",
        ),
        state: Exited,
    },
    ContainerProcess {
        id: ContainerId(
            "4133b65cffa94eb08caefba96e8182e119efccd9655d3230365bc73b0a6e109e",
        ),
        names: Name(
            "trusting_antonelli",
        ),
        state: Exited,
    },
]
